google-cloud-core/src/main/java/com/google/cloud/ServiceOptions.java

```diff
@@ -86,6 +86,8 @@
   private static final RetrySettings NO_RETRY_SETTINGS = getDefaultRetrySettingsBuilder()
       .setMaxAttempts(1).build();
 
+  private static final Pattern projectIdPattern = Pattern.compile("^[a-z][a-z0-9-]*[a-z0-9]+$");
```

Comment:
I think there is a bug here. `a-b-c` should be a valid project (I think) but the regex wouldn't match.

I think the right one would be `^[a-z][a-z0-9-]*$`

Reply:
Sorry it is. How about `^[a-z]([a-z0-9]*-*)*[a-z0-9]$`? `^[a-z][a-z0-9]*$` would match "abc-", but trailing hyphens are not allowed.

Reply:
Ah. Honestly regex is getting...difficult. Do you think we can just do this?

```java
for (char c : s.toCharArray()) {
  if (!isLetter(c) && !isDigit(c)) return false;
}
return s.length() > 0 && isLetter(s.charAt(0)) && !s.endsWith("-")
```

You'll need to implement isLetter and isDigit but those are easy.

Reply:
Yes. I agree with you on this. Will have a push soon